In [8]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.datasets import load_digits

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [9]:
#Loading the data
digits = load_digits()
digits

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [10]:
#dividing the data into X and y
X = digits.data
y = digits.target

In [11]:
#splitting the data into training and testing, default stratify
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.10, stratify=y)

In [12]:
# Runing multiple kernels and checking their classification report 
clf = KNeighborsClassifier()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(f"\t{clf.__class__.__name__}  :  Classification Report:")
print(classification_report(y_test, y_pred))

	KNeighborsClassifier  :  Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       160
           1       0.80      0.94      0.86       164
           2       0.94      0.94      0.94       159
           3       0.94      0.92      0.93       165
           4       0.99      0.91      0.95       163
           5       0.96      0.96      0.96       164
           6       0.96      0.98      0.97       163
           7       0.89      0.97      0.93       161
           8       0.90      0.78      0.83       157
           9       0.92      0.85      0.88       162

    accuracy                           0.93      1618
   macro avg       0.93      0.92      0.92      1618
weighted avg       0.93      0.93      0.93      1618



In [13]:
clf = DecisionTreeClassifier(splitter='best')
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(f"\t{clf.__class__.__name__}  :  Classification Report:")
print(classification_report(y_test, y_pred))

	DecisionTreeClassifier  :  Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.86      0.89       160
           1       0.66      0.63      0.64       164
           2       0.79      0.68      0.73       159
           3       0.73      0.81      0.77       165
           4       0.83      0.66      0.74       163
           5       0.59      0.58      0.59       164
           6       0.67      0.87      0.76       163
           7       0.78      0.76      0.77       161
           8       0.45      0.47      0.46       157
           9       0.63      0.67      0.65       162

    accuracy                           0.70      1618
   macro avg       0.71      0.70      0.70      1618
weighted avg       0.71      0.70      0.70      1618



In [16]:
clf = RandomForestClassifier(n_estimators=100)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(f"\t{clf.__class__.__name__}  :  Classification Report:")
print(classification_report(y_test, y_pred))

	RandomForestClassifier  :  Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.99      0.98       160
           1       0.83      0.91      0.87       164
           2       0.99      0.91      0.95       159
           3       0.93      0.90      0.92       165
           4       0.99      0.93      0.96       163
           5       0.96      0.93      0.95       164
           6       0.96      0.98      0.97       163
           7       0.94      0.89      0.91       161
           8       0.88      0.85      0.86       157
           9       0.80      0.91      0.85       162

    accuracy                           0.92      1618
   macro avg       0.92      0.92      0.92      1618
weighted avg       0.92      0.92      0.92      1618



In [17]:
# Runing multiple kernels and checking their classification report 
for clf in [ KNeighborsClassifier(), DecisionTreeClassifier(splitter='best'), RandomForestClassifier(n_estimators=100) ] :
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(f"\t{clf.__class__.__name__}  :  Classification Report:")
    print(classification_report(y_test, y_pred))

	KNeighborsClassifier  :  Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       160
           1       0.80      0.94      0.86       164
           2       0.94      0.94      0.94       159
           3       0.94      0.92      0.93       165
           4       0.99      0.91      0.95       163
           5       0.96      0.96      0.96       164
           6       0.96      0.98      0.97       163
           7       0.89      0.97      0.93       161
           8       0.90      0.78      0.83       157
           9       0.92      0.85      0.88       162

    accuracy                           0.93      1618
   macro avg       0.93      0.92      0.92      1618
weighted avg       0.93      0.93      0.93      1618

	DecisionTreeClassifier  :  Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.86      0.84       160
           1       0.68      0.6

## Conclusion
From the above classification report, we can observe that the classification accuracy were achieved by KNeighborsClassifier is 92% and by DecisionTreeClassifer is 74% while using 10% data for training and 90% dataset for testing. We can say that KNeighborsClassifier performs far better than Decision Tree Classifier in this case.